packages/react-native-web/src/modules/applyLayout/index.js

```diff
@@ -9,29 +9,73 @@
 
 import { canUseDOM } from 'fbjs/lib/ExecutionEnvironment';
 import debounce from 'debounce';
+import findNodeHandle from '../../exports/findNodeHandle';
 
 const emptyObject = {};
 const registry = {};
 
 let id = 1;
 const guid = () => `r-${id++}`;
 
+let resizeObserver;
 if (canUseDOM) {
-  const triggerAll = () => {
-    Object.keys(registry).forEach(key => {
-      const instance = registry[key];
-      instance._handleLayout();
+  if (typeof window.ResizeObserver !== 'undefined') {
+    resizeObserver = new window.ResizeObserver(entries => {
+      entries.forEach(({ target }) => {
+        target._handleLayout && target._handleLayout();
+      });
     });
-  };
+  } else {
+    if (process.env.NODE_ENV !== 'production') {
+      console.warn(
+        'onLayout relies on ResizeObserver which is not supported by your browser. ' +
+          'Please include a polyfill. https://github.com/WICG/ResizeObserver/issues/3. ' +
+          'Falling back to window.onresize.'
+      );
+    }
 
-  window.addEventListener('resize', debounce(triggerAll, 16), false);
+    const triggerAll = () => {
+      Object.keys(registry).forEach(key => {
+        const instance = registry[key];
+        instance._handleLayout();
+      });
+    };
+
+    window.addEventListener('resize', debounce(triggerAll, 16), false);
+  }
 }
 
+const observe = instance => {
+  if (resizeObserver) {
+    const node = findNodeHandle(instance);
+    node._handleLayout = debounce(instance._handleLayout.bind(instance));
```

Comment:
I'm applying this locally and noticed this. It would be good if we could avoid binding a new function to the DOM node and instead use a pointer (like `_onLayoutId`) to the instance's method. Did you try that? Also wondering why this function is debounced.

Reply:
Good point, `node._onLayoutId` and look up the instance in the `registry` should work.

RE: debouncing

As I said above
> Since the measure method you implemented takes this into account already, I decided to use ResizeObserver only to trigger instance._handleLayout calls. For this reason I am debouncing this function now because it triggers a new measurement which can be expensive (correct me if I am wrong here or think that we can or should not debounce).

Measurements will be triggered, whether the user callback is debunced/throttled or not. If we decide to get rid of the debouncing I think that, for the sake of consistency, we should also remove the debouncing here https://github.com/necolas/react-native-web/pull/848/files#diff-fa9d0b1cb86f4e38c30f6685ec719be4R44

Reply:
That's not debouncing measurements but the frequency with which onresize is called

Reply:
Right but as a consequence of that measurements (which happen in `_handleLayout`) will be delayed to when resize ends. 

> Also wondering why this function is debounced

That was to keep the behavior consistent otherwise, on resize, ResizeObserver would fire continuously and we'd invoke `_handleLayout` (measure) without delays.

Let me know if I am missing something.

Reply:
Hmm I see. I wonder if it's helpful to debounce at all. I suspect most viewport resizing is discrete - orientation change, etc - rather than the kind of continuous resizing developers might do. And maybe debouncing is better left to app developers who know when they want to use it? There's also the fact that measurement uses rAF, which needs to change too.

I'm working on a branch for FlatList that I'll push up so you can see how this patch interacts with it. I do want to get this in, it's just surfacing a few existing issues related to layout measurement in the lib :)

Reply:
> And maybe debouncing is better left to app developers who know when they want to use it?

Yes, that would be ideal. With the current implementation though they could only debounce `onLayout` (the fn prop), measurements would happen regardless. https://github.com/giuseppeg/react-native-web/blob/9db10f12c4dd23c46adb3f33f86932fd57f1f33a/packages/react-native-web/src/modules/applyLayout/index.js#L124-L132

Reply:
That's true. But probably ok. What I'll do is create a feature branch for this block of work, apply a version of this patch, and then we can work out these other details from there. Thanks for taking the time to understand (and help me understand) the details and for being patient!

Reply:
No problem! Sounds good, I pushed the change you suggested in https://github.com/necolas/react-native-web/pull/848#discussion_r178907038 I didn't bother rebasing but I can do that if you want.